includes/notes/class-wc-payments-notes-loan-approved.php

```diff
@@ -0,0 +1,165 @@
+		// If we have the correct information, proceed. Otherwise, delete existing notes.
+		if ( ! self::validate_inputs() ) {
+			// We don't have the necessary info to create a note, do nothing.
+			return;
+		}
+
+		// Check if the current loan info matches with the received one. If it matches, don't add a new one.
+		if ( ! self::check_attached_loan_data_is_different() ) {
+			// Loan paid out dates are the same, do nothing.
+			return;
+		}
+
+		// Do the overridden work.
+		if ( ! self::can_be_added() ) {
+			return;
+		}
+
+		$new_note = self::get_note();
+		$new_note->save();
+	}
+
+	/**
+	 * Check if the stored loan info has all the values we need.
+	 *
+	 * @return bool
+	 */
+	private static function validate_inputs() {
+		// If the loan amount isn't set correctly, don't push the note, and delete the old one if exists.
+		if ( empty( self::$loan_info )
+			|| ! is_array( self::$loan_info )
+			|| ! isset( self::$loan_info['details']['advance_amount'], self::$loan_info['details']['advance_paid_out_at'] )
+			|| ! is_numeric( self::$loan_info['details']['advance_amount'] )
+		) {
+			// There's something wrong with the loan information, delete the existing note, just in case of wrong information.
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Checks the saved paid out date on the previous note and deletes it if it doesn't match, to create a new one.
+	 *
+	 * @return bool
+	 */
+	private static function check_attached_loan_data_is_different() {
+		// Check if the note already exists, and the stored paid out date matches our current loan before adding a new one.
+		/**
+		 * Note class. Suppressed psalm error for WC<=5.5.0 because it uses an old class for the note.
+		 *
+		 * @var WC_Admin_Note|Note
+		 * @psalm-suppress UndefinedDocblockClass
+		 */
+		$note_class  = WC_Payment_Woo_Compat_Utils::get_note_class();
+		$notes_class = WC_Payment_Woo_Compat_Utils::get_notes_class();
+		$data_store  = WC_Data_Store::load( 'admin-note' );
+		$note_ids    = $data_store->get_notes_with_name( self::NOTE_NAME );
+
+		if ( ! empty( $note_ids ) ) {
+			$note = $notes_class::get_note( $note_ids[0] );
+			if ( $note instanceof $note_class ) {
+				$content_data = (array) $note->get_content_data();
+				if ( isset( $content_data['advance_paid_out_at'], $content_data['advance_amount'] ) ) {
+					if ( self::$loan_info['details']['advance_paid_out_at'] === $content_data['advance_paid_out_at'] &&
+						self::$loan_info['details']['advance_amount'] === $content_data['advance_amount'] ) {
+						// Note already exists for the current loan. No action will be taken.
+						return false;
+					}
+				}
+				// The note isn't for the current loan. Delete it to create a new one.
+				$data_store->delete( $note );
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Sets the loan information on the class.
+	 *
+	 * @param array $loan_info loan information.
+	 */
+	public static function set_loan_details( array $loan_info ) {
+		self::$loan_info = $loan_info;
```

Comment:
I don't see the point of this being a class variable. It's only being used on `possibly_add_note` (and functions called inside it). What do you think about passing the loan info (or individual properties) as parameters to `possibly_add_note` and other functions? I realize this is a matter of personal preference.

Reply:
Sure, that makes sense. I'll change it.

Reply:
Oh, now I realized why I did that in the first place, that the trait bound to this class expects `get_note` and `possibly_add_note` to be without arguments. So I'll keep this implementation.